README.md
# Demands

[![Build Status](https://travis-ci.org/yola/demands.png)](https://travis-ci.org/yola/demands)
[![Coverage Status](https://coveralls.io/repos/yola/demands/badge.png?branch=master)](https://coveralls.io/r/yola/demands?branch=master)
[![Latest Version](https://pypip.in/v/demands/badge.png)](https://pypi.python.org/pypi/demands/)
[![Downloads](https://pypip.in/d/demands/badge.png)](https://pypi.python.org/pypi/demands/)

A base HTTP service client.

By default it "demands" successful responses from API endpoints,
otherwise it raises an exception.

Demands accepts all the same parameters as `requests.request` and extends
the `requests.Session` class, documentation for both:
[Requests Developer Interface][2].

Written and used by the folks at Yola to support our [free website builder][1].

## HTTPServiceClient Overview

* base class for creating service clients
* provides flexible way of http error handling for descendants
* `HTTPServiceError` raised on unexpected service response

* Supports pre-send and post-send hooks

## Usage
```python
from demands import HTTPServiceClient

class MyService(HTTPServiceClient):
    def get_user(self, user_id):
        return self.get('/users/%s/' % user_id).json()

    def safe_get_user(self, user_id, default_user):
        response = self.get(
            '/users/%s/' % user_id, 
            expected_response_codes=[404])
        return response.json() if response.is_ok else default_user


service = MyService(url='http://localhost/')
user = service.get_user(1234)
```

Any parameters passed to the constructor will also be used for 
each and every request:
```python
service = MyService(
    url='http://localhost/',
    headers={'h1':'value'},
    auth=('username','pass'),
)

# sent with auth and both headers
user = service.get('/some-path', headers={'h2': 'kittens'})
```


## Testing

Install development requirements:

    pip install -r requirements.txt

Run the tests with:

    nosetests

## API documentation

To generate the documentation:

    cd docs && PYTHONPATH=.. make singlehtml

[1]:https://www.yola.com/
[2]:http://www.python-requests.org/en/latest/api/
[3]:https://github.com/kennethreitz/requests


CHANGELOG.md
# [Changelog](https://github.com/yola/demands/releases)

## 5.1.0

* Fix wrong behavior of `Page.is_last_page`
* Remove `testtube`

## 5.0.0

* Remove deprecation warning in Python 3.
* Add Python 3.5, 3.6, 3.7, 3.8 as supported.
* Stop supporting Python 3.3 and 3.4.

## 4.1.0

* Add proper handling of empty next page in paginated result set.
* PaginatedResults now accepts `next_key` option.

## 4.0.0

* Rename `PaginatedAPIIterator` to `PaginatedResults`.
* Remove JSONServiceClient. Newest `requests` version provides functionality
  that obviates the need for it.

## 3.0.0

* Add decoding of json responses to JSONServiceClient

## 2.0.0

* Add JSONServiceClient
* Remove `send_as_json` configuration for HTTPServiceClient
* Changes to PaginatedAPIIterator:
    * Default start page is 1 for page-based pagination (was 0)
    * Expects the paginated function to return a dict with a `'results'` key
    * See PaginatedAPIIterator docstring for customizing those behaviors

## 1.3.1

* Add MANIFEST.in

## 1.3.0 - Broken

* Add PaginatedAPIIterator, helper utility for paginated service methods

## 1.2.0

* Allow the user to customise the definition of an acceptable response by overriding `is_acceptable`.
* Exceptions are now raised directly by `request` if `is_acceptable()` returns `False` instead of in the default
`post_send`.

## 1.1.5

* No change: a broken version of 1.1.4 was released accidentally (files were missing from the package).

## 1.1.4

* Stop logging an error for unsuccessful responses. Move the information
  from the logging statement to the exception message.

## 1.1.3

* Make Demands work correctly if `path` param is empty. Don't add slash to
  the base URL in this case.

## 1.1.0

* Add Python 3 support

## 1.0.6

* Switch to version `requests` >= `2.4.2`

## 1.0.5

* Allows to set max retries

## 1.0.4

* Accept param `verify_ssl` as an alias for `verify`
* Fix easy_install error, a pitfall of using `__init__.py`

## 1.0.3

* Adds API documentation
* Add a license

## 1.0.2

* Adds http response as property of HTTPServiceError exception
* HTTPServiceError now subclasses AssertionError instead of IOError

## 1.0.1

* Fixes issue with composing url

## 1.0.0

* HTTPServiceClient invoked with arguments instead of a dictionary
* Enable and match parameters used by `requests`

    Before
    * `data`, used for both query string and request body
    * `verify_ssl`

    Now
    * `params`: dictionary or bytes to be sent in the query string
    * `data`: dictionary, bytes, or file-like object to send in the body
    * `verify`: verify the SSL cert

## 0.2.1

* Save error response content in HTTPServiceError exception

## 0.2.0

* Switches to requests > 1.0.0

## 0.1.0

* Adds client identification
* Provides an easier way to modify request arguments in pre_send()

## 0.0.1

* Initial version


tests/test_models.py
import unittest2
import inspect

from requests import Session, Response
from mock import Mock, patch

from demands.models import HTTPServiceClient, HTTPServiceError


class PatchedSessionTests(unittest2.TestCase):
    def setUp(self):
        # must patch inspect since it is used on Session.request, and when
        # Session.request is mocked, inspect blows up
        self.request_args = inspect.getargspec(Session.request)
        self.inspect_patcher = patch('demands.models.inspect.getargspec')
        self.patched_inspect = self.inspect_patcher.start()
        self.patched_inspect.return_value = self.request_args

        self.request_patcher = patch.object(Session, 'request')
        self.request = self.request_patcher.start()
        self.response = Mock(spec=Response(), status_code=200)
        self.request.return_value = self.response

    def tearDown(self):
        self.request_patcher.stop()
        self.inspect_patcher.stop()


class HttpServiceTests(PatchedSessionTests):

    def setUp(self):
        PatchedSessionTests.setUp(self)
        self.service = HTTPServiceClient('http://service.com/')

    def test_returning_responses_from_all_session_calls(self):
        self.assertEqual(self.service.get('/path'), self.response)
        self.assertEqual(self.service.put('/path'), self.response)
        self.assertEqual(self.service.delete('/path'), self.response)
        self.assertEqual(self.service.post('/path'), self.response)
        self.assertEqual(self.service.patch('/path'), self.response)
        self.assertEqual(self.service.options('/path'), self.response)
        self.assertEqual(self.service.head('/path'), self.response)

    def test_get_request_with_params(self):
        """GET request with url parameters"""
        self.service.get('/get-endpoint', params={'foo': 'bar'})
        self.request.assert_called_with(
            method='GET', url='http://service.com/get-endpoint',
            allow_redirects=True, params={'foo': 'bar'})

    def test_minimal_post_request(self):
        """minimal POST request"""
        self.service.post('/post-endpoint')
        self.request.assert_called_with(
            method='POST', url='http://service.com/post-endpoint', data=None)

    def test_minimal_put_request(self):
        """minimal PUT request"""
        self.service.put('/put-endpoint')
        self.request.assert_called_with(
            method='PUT', url='http://service.com/put-endpoint', data=None)

    def test_minimal_delete_request(self):
        """minimal DELETE request"""
        self.service.delete('/delete-endpoint')
        self.request.assert_called_with(
            method='DELETE', url='http://service.com/delete-endpoint')

    def test_sets_authentication_when_username_provided(self):
        service = HTTPServiceClient(
            url='http://localhost/',
            username='foo',
            password='bar',
        )
        service.get('/authed-endpoint')
        self.request.assert_called_with(
            method='GET', url='http://localhost/authed-endpoint',
            allow_redirects=True, auth=('foo', 'bar'))

    def test_client_identification_adds_user_agent_header(self):
        """client identification adds User-Agent header"""
        service = HTTPServiceClient(
            url='http://localhost/',
            client_name='my_client',
            client_version='1.2.3',
            app_name='my_app',
            headers={'Foo': 'Bar'},
        )
        service.get('/test')
        self.request.assert_called_with(
            method='GET', url='http://localhost/test', allow_redirects=True,
            headers={'User-Agent': 'my_client 1.2.3 - my_app', 'Foo': 'Bar'})

    def test_post_send_raise_exception_in_case_of_error(self):
        self.response.configure_mock(url='http://broken/', status_code=500)
        with self.assertRaises(HTTPServiceError):
            self.service.request('METHOD', 'http://broken/')

    def test_post_send_raises_exception_with_details_on_error(self):
        self.response.configure_mock(
            status_code=500, content='content', url='http://broken/')
        with self.assertRaises(HTTPServiceError) as e:
            self.service.request('METHOD', 'http://broken/')
            self.assertEqual(e.exception.code, 500)
            self.assertEqual(e.exception.details, 'content')

    def test_post_sends_no_exception_in_case_of_expected_response_code(self):
        self.response.configure_mock(
            status_code=404, content='content', url='http://notfound/')
        self.service.request(
            'METHOD', 'http://notfound/', expected_response_codes=(404,))

    @patch('demands.models.log')
    def test_post_send_logs_errors(self, mock_log):
        """failed requests are logged with status code, and content"""
        self.response.configure_mock(
            status_code=500, content='content', url='http://service.com/')
        with self.assertRaises(HTTPServiceError):
            self.service.request('METHOD', 'http://service.com/')
            error_msg = get_parsed_log_message(mock_log, 'error')
            self.assertIn('service.com', error_msg)
            self.assertIn('500', error_msg)
            self.assertIn('content', error_msg)

    def test_santization_of_request_parameters_removes_unknowns(self):
        lots_of_params = {
            'expected_response_codes': (404,),
            'method': 'METHOD',
            'something_odd': True
        }
        self.assertEqual(
            self.service._sanitize_request_params(lots_of_params),
            {'method': 'METHOD'})


def get_parsed_log_message(mock_log, log_level):
    """Return the parsed log message sent to a mock log call at log_level

    Example:

        log = Mock()

        def do_something(data1, data2):
            log.info('Doing something with %s and %s', data1, data2)

        do_something('one', 'two')

        get_parsed_log_message(log, 'error')
        # => 'Doing something with one and two'

    """
    call_args = getattr(mock_log, log_level).call_args
    if not call_args:
        raise Exception('%s.%s was not called' % (mock_log, log_level))
    args, kwargs = call_args
    return args[0] % tuple(args[1:])


tests/test_demands.py
# -*- coding: utf-8 -*-
import inspect
import json
from unittest import TestCase

from requests import Session, Response
from mock import Mock, patch
from six import itervalues

from demands import HTTPServiceClient, HTTPServiceError


class PatchedSessionTests(TestCase):
    def setUp(self):
        # must patch inspect since it is used on Session.request, and when
        # Session.request is mocked, inspect blows up
        self.request_args = inspect.getargspec(Session.request)
        self.inspect_patcher = patch('demands.inspect.getargspec')
        self.patched_inspect = self.inspect_patcher.start()
        self.patched_inspect.return_value = self.request_args

        self.request_patcher = patch.object(Session, 'request')
        self.request = self.request_patcher.start()
        self.response = Mock(spec=Response(), status_code=200)
        self.request.return_value = self.response

    def tearDown(self):
        self.request_patcher.stop()
        self.inspect_patcher.stop()


class HttpServiceTests(PatchedSessionTests):

    def setUp(self):
        PatchedSessionTests.setUp(self)
        self.service = HTTPServiceClient('http://service.com/')

    def test_returning_responses_from_all_session_calls(self):
        self.assertEqual(self.service.get('/path'), self.response)
        self.assertEqual(self.service.put('/path'), self.response)
        self.assertEqual(self.service.delete('/path'), self.response)
        self.assertEqual(self.service.post('/path'), self.response)
        self.assertEqual(self.service.patch('/path'), self.response)
        self.assertEqual(self.service.options('/path'), self.response)
        self.assertEqual(self.service.head('/path'), self.response)

    def test_get_request_with_params(self):
        """GET request with url parameters"""
        self.service.get('/get-endpoint', params={'foo': 'bar'})
        self.request.assert_called_with(
            method='GET', url='http://service.com/get-endpoint',
            allow_redirects=True, params={'foo': 'bar'})

    def test_minimal_post_request(self):
        """minimal POST request"""
        self.service.post('/post-endpoint')
        self.request.assert_called_with(
            method='POST', url='http://service.com/post-endpoint', json=None,
            data=None)

    def test_minimal_put_request(self):
        """minimal PUT request"""
        self.service.put('/put-endpoint')
        self.request.assert_called_with(
            method='PUT', url='http://service.com/put-endpoint', data=None)

    def test_minimal_delete_request(self):
        """minimal DELETE request"""
        self.service.delete('/delete-endpoint')
        self.request.assert_called_with(
            method='DELETE', url='http://service.com/delete-endpoint')

    def test_unacceptable_response(self):
        def get():
            self.service.get('/get-endpoint')
        acceptable = True
        self.service.is_acceptable = lambda *args: acceptable
        get()
        acceptable = False
        self.assertRaises(HTTPServiceError, get)

    def test_headers_are_passed_and_overridable(self):
        service = HTTPServiceClient(
            url='http://localhost/',
            headers={
                'name': 'value',
                'thomas': 'kittens'})
        service.get('/')
        self.request.assert_called_with(
            headers={'thomas': 'kittens', 'name': 'value'},
            method='GET', url='http://localhost/', allow_redirects=True)
        service.get('/', headers={'thomas': 'homegirl'})
        self.request.assert_called_with(
            headers={'thomas': 'homegirl', 'name': 'value'},
            method='GET', url='http://localhost/', allow_redirects=True)

    def test_sets_authentication_when_provided(self):
        service = HTTPServiceClient(
            url='http://localhost/',
            auth=('foo', 'bar'),
        )
        service.get('/authed-endpoint')
        self.request.assert_called_with(
            method='GET', url='http://localhost/authed-endpoint',
            allow_redirects=True, auth=('foo', 'bar'))

    @patch('demands.log')
    def test_logs_authentication_when_provided(self, mock_log):
        service = HTTPServiceClient(
            url='http://localhost/',
            auth=('foo', 'bar'),
        )
        service.get('/authed-endpoint')
        debug_msgs = get_parsed_log_messages(mock_log, 'debug')
        self.assertIn('Authentication', debug_msgs[2])

    def test_client_identification_adds_user_agent_header(self):
        """client identification adds User-Agent header"""
        service = HTTPServiceClient(
            url='http://localhost/',
            client_name='my_client',
            client_version='1.2.3',
            app_name='my_app',
            headers={'Foo': 'Bar'},
        )
        service.get('/test')
        self.request.assert_called_with(
            method='GET', url='http://localhost/test', allow_redirects=True,
            headers={'User-Agent': 'my_client 1.2.3 - my_app', 'Foo': 'Bar'})

    def test_post_send_raise_exception_in_case_of_error(self):
        self.response.configure_mock(url='http://broken/', status_code=500)
        with self.assertRaises(HTTPServiceError):
            self.service.request('METHOD', 'http://broken/')

    def test_post_send_raises_exception_with_details_on_error(self):
        self.response.configure_mock(
            status_code=500, content='content', url='http://broken/')
        with self.assertRaises(HTTPServiceError) as e:
            self.service.request('METHOD', 'http://broken/')
            self.assertEqual(e.exception.code, 500)
            self.assertEqual(e.exception.details, 'content')
            self.assertTrue(e.message.startswith(
                'Unexpected response: url: '
                'http://broken/, code: 500, details: '
            ))

    def test_post_sends_no_exception_in_case_of_expected_response_code(self):
        self.response.configure_mock(
            status_code=404, content='content', url='http://notfound/')
        self.service.request(
            'METHOD', 'http://notfound/', expected_response_codes=(404,))

    def test_santization_of_request_parameters_removes_unknowns(self):
        lots_of_params = {
            'expected_response_codes': (404,),
            'method': 'METHOD',
            'something_odd': True
        }
        self.assertEqual(
            self.service._sanitize_request_params(lots_of_params),
            {'method': 'METHOD'})

    def test_santization_of_verify_parameters(self):
        lots_of_params = {
            'expected_response_codes': (404,),
            'method': 'METHOD',
            'verify_ssl': '/some/path',
        }
        self.assertEqual(
            self.service._sanitize_request_params(lots_of_params),
            {'method': 'METHOD', 'verify': '/some/path'})

    def test_url_is_composed_properly(self):
        service = HTTPServiceClient('http://service.com/some/path/')
        service.get('/get-endpoint')
        self.request.assert_called_with(
            method='GET', url='http://service.com/some/path/get-endpoint',
            allow_redirects=True
        )

    def test_url_is_composed_properly_if_path_is_empty(self):
        service = HTTPServiceClient(
            'http://service.com/some/path/get-endpoint')
        service.get('')
        self.request.assert_called_with(
            method='GET', url='http://service.com/some/path/get-endpoint',
            allow_redirects=True
        )

    def test_pre_send_sets_max_retries(self):
        self.service.pre_send({'max_retries': 2})
        for adapter in itervalues(self.service.adapters):
            self.assertEqual(adapter.max_retries, 2)

    def test_pre_send_defaults_max_retries_to_zero(self):
        self.service.pre_send({'max_retries': 2})
        self.service.pre_send({})
        for adapter in itervalues(self.service.adapters):
            self.assertEqual(adapter.max_retries, 0)


def get_parsed_log_messages(mock_log, log_level):
    """Return the parsed log message sent to a mock log call at log_level

    Example:

        log = Mock()

        def do_something(data1, data2):
            log.info('Doing something with %s and %s', data1, data2)

        do_something('one', 'two')

        get_parsed_log_message(log, 'error')
        # => 'Doing something with one and two'

    """
    calls = getattr(mock_log, log_level).call_args_list
    if not calls:
        raise Exception('%s.%s was not called' % (mock_log, log_level))
    messages = []
    for call_args in calls:
        args, kwargs = call_args
        messages.append(args[0] % tuple(args[1:]))
    return messages


tests/test_service.py
import unittest2

import requests
from mock import Mock, patch

from demands.service import Request, HTTPService, HTTPServiceError


class PatchedRequestsTests(unittest2.TestCase):
    def setUp(self):
        self.requests_patcher = patch('demands.service.requests')
        self.requests = self.requests_patcher.start()
        self.response = Mock(spec=requests.Response, status_code=200,
                             content='', headers={'content-type': 'application/json'})

        for method in ('get', 'post', 'put', 'patch', 'delete'):
            fn = getattr(self.requests, method)
            fn.return_value = self.response

    def tearDown(self):
        self.requests_patcher.stop()


class RequestTests(PatchedRequestsTests):
    url = 'http://localhost/'

    def test_request_properly_authenticates(self):
        username, password = 'stevemcqueen', 'password'
        url = 'https://localhost/'
        request = Request(url, 'POST')
        request.authenticate(username, password)
        request.send()
        self.requests.post.assert_called_once_with(url, auth=(username, password),
                                                   verify=True)

    def test_request_sends_proper_arguments_for_headers_cookies_and_data(self):
        # Simple request
        request = Request(self.url, 'POST')
        request.send()
        self.requests.post.assert_called_with(self.url, auth=None)

        # Request with headers
        headers = {'Content-Type': 'text/html'}
        request = Request(self.url, 'GET', headers=headers)
        request.send()
        self.requests.get.assert_called_with(self.url, headers=headers, auth=None)

        # Request with data and cookies
        data = {'robots': 1}
        cookies = {'cookie_name': 'cookie_value'}
        request = Request(self.url, 'POST', data=data, cookies=cookies)
        request.send()
        self.requests.post.assert_called_with(self.url, data=data, cookies=cookies, auth=None)


class HttpServiceRequestTests(PatchedRequestsTests):
    def test_request_is_sent_with_correct_headers(self):
        url = 'http://localhost/'
        path = 'path'
        headers = {'its simple': 'we kill the batman'}
        service = HTTPService({
            'url': url,
        })
        service.get(path, headers=headers)
        self.requests.get.assert_called_with(url + path, params={},
                                             headers=headers, verify=True,
                                             auth=None)

    def test_request_is_sent_with_correct_cookies(self):
        url = 'http://localhost/'
        path = 'path'
        cookies = {'its simple': 'we kill the batman'}
        service = HTTPService({
            'url': url,
        })
        service.get(path, cookies=cookies)
        self.requests.get.assert_called_with(url + path, params={},
                                             cookies=cookies, verify=True,
                                             auth=None)

class HttpServiceTests(unittest2.TestCase):
    def setUp(self):
        self.service = HTTPService({
            'url': 'http://service.com/'
        })
        self.request = Request('http://service.com/', 'GET')
        self.response = Mock(headers={'content-type': 'text/plan'})

    def test_pre_send_triggers_authentication_when_username_provided(self):
        service = HTTPService({
            'url': 'http://localhost/',
            'username': 'foo',
            'password': 'bar'
        })
        service.pre_send(self.request)
        self.assertEqual(self.request.auth, ('foo', 'bar'))

    def test_pre_send_can_trigger_client_identification(self):
        service = HTTPService({
            'url': 'http://localhost/',
            'client_name': 'my_client',
            'client_version': '1.2.3',
            'app_name': 'my_app'
        })
        service.pre_send(self.request)
        self.assertEqual(self.request.arguments['headers']['User-Agent'],
                         'my_client 1.2.3 - my_app')

    def test_post_send_raises_exception_in_case_of_error(self):
        self.response.configure_mock(status_code=500, content='content')
        self.assertRaises(HTTPServiceError, self.service.post_send, self.request, self.response)

    def test_post_send_raises_exception_with_details_on_error(self):
        self.response.configure_mock(status_code=500, content='content')
        try:
            self.service.post_send(self.request, self.response)
            self.assertFail('HTTPServiceError not raised')
        except HTTPServiceError, e:
            self.assertEqual(e.code, 500)
            self.assertEqual(e.details, 'content')

    def test_post_send_do_not_raise_exception_in_case_of_expected_response_code(self):
        self.response.configure_mock(status_code=404, content='content')
        self.service.post_send(self.request, self.response, expected_response_codes=(404,))

    @patch('demands.service.log')
    def test_post_send_logs_errors(self, mock_log):
        """failed requests are logged with status code, and content"""
        self.response.configure_mock(status_code=500, content='content')
        with self.assertRaises(HTTPServiceError):
            self.service.post_send(self.request, self.response)
            error_msg = get_parsed_log_message(mock_log, 'error')
            self.assertIn('service.com', error_msg)
            self.assertIn('500', error_msg)
            self.assertIn('content', error_msg)


def get_parsed_log_message(mock_log, log_level):
    """Return the parsed log message sent to a mock log call at log_level

    Example:

        log = Mock()

        def do_something(data1, data2):
            log.info('Doing something with %s and %s', data1, data2)

        do_something('one', 'two')

        get_parsed_log_message(log, 'error')
        # => 'Doing something with one and two'

    """
    call_args = getattr(mock_log, log_level).call_args
    if not call_args:
        raise Exception('%s.%s was not called' % (mock_log, log_level))
    args, kwargs = call_args
    return args[0] % tuple(args[1:])


demands/models.py
import logging
import time
import inspect
import copy

from urlparse import urljoin
from requests import Session


log = logging.getLogger(__name__)


class HTTPServiceError(Exception):
    def __init__(self, response):
        try:
            self.details = response.json()
        except ValueError:
            self.details = response.content
        super(Exception, self).__init__(
            'code: %s, details: %s' % (response.status_code, self.details)
        )


class HTTPServiceClient(Session):
    """Extendable base service client object"""

    def __init__(self, url, **kwargs):
        super(HTTPServiceClient, self).__init__()
        self.url = url
        self._shared_request_params = {}
        self._shared_request_params = self._get_request_params(**kwargs)

        if 'client_name' in kwargs:
            # client name and version is important because we want to
            # accurately log errors and throw deprecation warns when outdated
            headers = self._shared_request_params.get('headers') or {}
            headers.update({
                'User-Agent': '%s %s - %s' % (
                    kwargs.pop('client_name'),
                    kwargs.pop('client_version', 'x.y.z'),
                    kwargs.pop('app_name', 'unknown'),
                )
            })
            self._shared_request_params['headers'] = headers

    def _get_request_params(self, **kwargs):
        """Return a copy of self._shared_request_params updated with kwargs"""
        request_params = copy.deepcopy(self._shared_request_params)

        if 'username' in kwargs:
            username = kwargs.pop('username')
            request_params['auth'] = (username, kwargs.pop('password', None))
            log.debug('Authentication via HTTP auth as "%s"', username)

        request_params.update(kwargs)
        return request_params

    def _sanitize_request_params(self, request_params):
        """Remove keyword arguments not used by `requests`"""
        valid_args = inspect.getargspec(Session.request)[0]
        return dict((key, val) for key, val in request_params.items()
                    if key in valid_args)

    def request(self, method, path, **kwargs):
        """"Configure params. Send a <Request>. Demand and return a <Response>.

        In addition to parameters allowed by `reqeust` there are:
        :param expected_response_codes: workaround for services which returns
            non-expected results, like we search for users - and expect []
            in case nobody is found, but got 404 instead.
        :param username: enables authenticated requests
        :param password: used in conjunction with username

        """
        url = urljoin(self.url, path)
        request_params = self._get_request_params(
            url=url, method=method, **kwargs)
        request_params = self.pre_send(request_params)
        request_params = (request_params)

        sanitized_params = self._sanitize_request_params(request_params)
        start_time = time.time()
        response = super(HTTPServiceClient, self).request(**sanitized_params)
        log.debug(
            '%s HTTP [%s] call to "%s" %.2fms',
            response.status_code, method, response.url,
            (time.time() - start_time) * 1000)
        log.debug('HTTP request params: %s', request_params)

        response = self.post_send(response, request_params=request_params)
        return response

    def pre_send(self, request_params):
        """"Override this method to modify sent request parameters"""
        return request_params

    def post_send(self, response, request_params={}):
        """"Override this method to modify returned response"""
        self._demand_success(response, request_params)
        return response

    def _demand_success(self, response, request_params):
        expected_codes = request_params.get('expected_response_codes', [])
        response.is_ok = response.status_code < 300
        if not (response.is_ok or response.status_code in expected_codes):
            log.error(
                'Unexpected response from %s: url: %s, code: %s, details: %s',
                self.__class__.__name__, response.url, response.status_code,
                response.content)
            raise HTTPServiceError(response)


demands/service.py
import logging
import requests
import time


log = logging.getLogger(__name__)


class Request(object):
    """
    Request object which provides functionality to send and receive
    http requests/responses.
    """

    def __init__(self, url, method, **kwargs):
        self.url = url
        self.method = method
        self.arguments = kwargs
        self.verify = kwargs.get('verify', True)
        self.auth = None

    def _compose_request_arguments(self):
        arguments = self.arguments
        if self.method not in ('POST', 'PUT', 'PATCH') and 'data' in arguments:
            arguments['params'] = arguments.pop('data')

        arguments['auth'] = self.auth

        if self.url.startswith('https'):
            arguments['verify'] = self.verify

        return arguments

    def authenticate(self, username, password):
        """Enable http authentication with the provided username and password"""
        log.debug('Authentication via HTTP auth as "%s"', username)
        self.auth = (username, password)

    def set_header(self, header, value):
        self.arguments.setdefault('headers', {})[header] = value

    def send(self):
        """Execute the request, and return the response"""
        method = self.method.lower()
        request_arguments = self._compose_request_arguments()
        start_time = time.time()
        response = getattr(requests, method)(self.url, **request_arguments)
        log.debug('%s HTTP [%s] call to "%s" %.2fms', response.status_code, self.method, self.url,
                  (time.time() - start_time) * 1000)
        log.debug('HTTP request data: %s', request_arguments)
        return response


class HTTPServiceError(Exception):
    def __init__(self, code, details):
        self.code = code
        self.details = details
        super(Exception, self).__init__(
            'code: %s, details: %s' % (code, details)
        )


class HTTPService(object):
    """
    Provides an interface which allows arbitrary methods to be defined and
    called on a remote http service.
    """
    def __init__(self, config):
        self.config = config

    def pre_send(self, request, **params):
        """
        Override to modify request object to be called just before sending
        the request
        """
        if 'username' in self.config:
            request.authenticate(
                self.config['username'], self.config['password'])

        if 'client_name' in self.config:
            request.set_header('User-Agent', '%s %s - %s' % (
                self.config['client_name'],
                self.config.get('client_version', 'x.y.z'),
                self.config.get('app_name', 'unknown'),
            ))

    def post_send(self, request, response, **params):
        """Override to modify response object returned by call made by request object."""
        response.is_ok = response.status_code < 300
        if (not response.is_ok and
                not response.status_code in params.get('expected_response_codes', [])):
            log.error('Unexpected response from %s: url: %s, code: %s, details: %s',
                self.__class__.__name__, response.url, response.status_code, response.content)

            raise HTTPServiceError(response.status_code, response.content)

    def get(self, path, **kwargs):
        return self._make_call('GET', path, **kwargs)

    def post(self, path, **kwargs):
        return self._make_call('POST', path, **kwargs)

    def put(self, path, **kwargs):
        return self._make_call('PUT', path, **kwargs)

    def delete(self, path, **kwargs):
        return self._make_call('DELETE', path, **kwargs)

    def _make_call(self, method, path, **kwargs):
        """
        Call the service method defined by the passed path and http method.
        Additional arguments include cookies, headers, and data values.
        """
        base = self.config.get('url')
        url = '/'.join([base.rstrip('/'), path.lstrip('/')])

        kwargs.setdefault('verify', self.config.get('verify_ssl', True))
        request = Request(url, method, **kwargs)

        self.pre_send(request, **kwargs)
        response = request.send()
        self.post_send(request, response, **kwargs)
        return response
